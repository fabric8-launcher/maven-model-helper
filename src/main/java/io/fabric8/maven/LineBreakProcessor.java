package io.fabric8.maven;

import org.jdom2.Element;
import org.jdom2.output.support.AbstractXMLOutputProcessor;
import org.jdom2.output.support.FormatStack;
import org.jdom2.util.NamespaceStack;

import java.io.IOException;
import java.io.Writer;
import java.util.List;

/**
 * Inserts a line break on every major section
 */
class LineBreakProcessor extends AbstractXMLOutputProcessor {

    @Override
    protected void printElement(Writer out, FormatStack fstack, NamespaceStack nstack, Element element) throws IOException {
        Element rootElement = element.getDocument().getRootElement();
        // Test if it's a major section
        int idxElement = rootElement.indexOf(element);
        List<Element> children = element.getChildren();
        boolean addBreak = (idxElement > -1 && !children.isEmpty());
        super.printElement(out, fstack, nstack, element);
        if (addBreak) {
            out.write(fstack.getLineSeparator());
        }
    }
}
