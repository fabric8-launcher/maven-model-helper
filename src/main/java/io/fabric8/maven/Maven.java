package io.fabric8.maven;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.Reader;
import java.io.StringReader;
import java.io.StringWriter;
import java.io.UncheckedIOException;
import java.io.Writer;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.function.Supplier;

import org.apache.maven.model.Model;
import org.apache.maven.model.io.xpp3.MavenXpp3Reader;
import org.apache.maven.model.io.xpp3.MavenXpp3Writer;
import org.codehaus.plexus.util.ReaderFactory;
import org.codehaus.plexus.util.xml.XmlStreamReader;
import org.codehaus.plexus.util.xml.pull.XmlPullParserException;
import org.jdom2.Document;
import org.jdom2.JDOMException;
import org.jdom2.input.SAXBuilder;
import org.jdom2.output.XMLOutputter;

/**
 * @author <a href="mailto:ggastald@redhat.com">George Gastaldi</a>
 */
public final class Maven {

    private Maven() {
        throw new IllegalAccessError("Utility class");
    }

    /**
     * Create a new Maven {@link Model}
     *
     * @return a new {@link Model}
     */
    public static Model newModel() {
        Model model = new Model();
        model.setModelVersion("4.0.0");
        model.setProperties(new SortedProperties());
        return model;
    }

    /**
     * A String version that makes it easier to be called from other languages
     *
     * @see #readModel(Path)
     */
    public static Model readModel(String path) {
        return readModel(Paths.get(path));
    }

    /**
     * Read the {@link Path} as a {@link Model}
     *
     * @param pom a path to a pom.xml file
     * @return the maven {@link Model}
     */
    public static Model readModel(Path pom) {
        try (BufferedReader br = Files.newBufferedReader(pom)) {
            Model model = readModel(br);
            model.setPomFile(pom.toFile());
            return model;
        } catch (IOException io) {
            throw new UncheckedIOException("Error while reading pom.xml", io);
        }
    }

    /**
     * Read the {@link Path} as a {@link Model}
     *
     * @param rdr a Reader on the contents of a pom file
     * @return the maven {@link Model}
     */
    public static Model readModel(Reader rdr) {
        try (Reader reader = rdr) {
            MavenXpp3Reader mavenXpp3Reader = new MavenXpp3Reader();
            Model model = mavenXpp3Reader.read(reader);
            // https://github.com/fabric8-launcher/maven-model-helper/issues/44
            SortedProperties sortedProps = new SortedProperties();
            sortedProps.putAll(model.getProperties());
            model.setProperties(sortedProps);
            return model;
        } catch (IOException io) {
            throw new UncheckedIOException("Error while reading pom.xml", io);
        } catch (XmlPullParserException e) {
            throw new RuntimeException("Error while parsing pom.xml", e);
        }
    }

    /**
     * Read the {@link InputStream} as a {@link Model}
     *
     * @param inputStream an input stream of a pom.xml file
     * @return the maven {@link Model}
     */
    public static Model readModel(InputStream inputStream) {
        try (XmlStreamReader xmlStreamReader = ReaderFactory.newXmlReader(inputStream)) {
            return readModel(xmlStreamReader);
        } catch (IOException io) {
            throw new UncheckedIOException("Error while reading stream", io);
        }
    }

    /**
     * Shortcut to writeModel(model,model.getPomFile().toPath());
     *
     * @param model the model to write
     */
    public static void writeModel(Model model) {
        writeModel(model, model.getPomFile().toPath());
    }

    /**
     * Write the Model back to the provided {@link Path}
     *
     * @param model the model to write
     * @param pom the path to the POM file
     */
    public static void writeModel(Model model, Path pom) {
        writeModel(model, pom, () -> {
            try {
                return Files.newBufferedWriter(pom);
            } catch (IOException e) {
                throw new UncheckedIOException("Could not write to Writer", e);
            }
        });
    }

    /**
     * Write the Model back to the provided {@link Path} using the specified {@link XMLFormat}
     *
     * @param model the model to write
     * @param pom the path to the POM file
     * @param format the XML format to use
     */
    public static void writeModel(Model model, Path pom, XMLFormat format) {
        writeModel(model, pom, () -> {
            try {
                return Files.newBufferedWriter(pom);
            } catch (IOException e) {
                throw new UncheckedIOException("Could not write to Writer", e);
            }
        }, format);
    }

    /**
     * Shortcut to writeModel(model,model.getPomFile().toPath(),writer);
     *
     * @param model the model to write
     * @param writer the writer to write the model to
     */
    public static void writeModel(Model model, Writer writer) {
        writeModel(model, model.getPomFile() != null ? model.getPomFile().toPath() : null, () -> writer);
    }

    /**
     * Shortcut to writeModel(model,model.getPomFile().toPath(),writer, format);
     *
     * @param model the model to write
     * @param writer the writer to write the model to
     * @param format the XML format to use
     */
    public static void writeModel(Model model, Writer writer, XMLFormat format) {
        writeModel(model, model.getPomFile() != null ? model.getPomFile().toPath() : null, () -> writer, format);
    }

    /**
     * Write the Model to the {@link Writer} using the provided {@link Path} as a reference
     *
     * @param model the model to write
     * @param pom the path to the POM file
     * @param writer the writer to write the model to
     */
    public static void writeModel(Model model, Path pom, Writer writer) {
        writeModel(model, pom, () -> writer);
    }

    /**
     * Write the Model to the {@link Writer} using the provided {@link Path} as a reference
     *
     * @param model the model to write
     * @param pom the path to the POM file
     * @param writerSupplier the writer supplier to write the model to
     */
    public static void writeModel(Model model, Path pom, Supplier<Writer> writerSupplier) {
        writeModel(model, pom, writerSupplier, null);
    }

    /**
     * Write the Model to the {@link Writer} using the provided {@link Path} as a reference
     *
     * @param model the model to write
     * @param pom the path to the POM file
     * @param writerSupplier the writer supplier to write the model to
     */
    public static void writeModel(Model model, Path pom, Supplier<Writer> writerSupplier, XMLFormat format) {
        if (pom == null || pom.toFile().length() == 0L) {
            // Initialize an empty XML
            try (Writer writer = writerSupplier.get()) {
                if (format != null) {
                    // Format specified, write to a String first
                    StringWriter sw = new StringWriter();
                    MavenXpp3Writer mavenXpp3Writer = new MavenXpp3Writer();
                    mavenXpp3Writer.write(sw, model);
                    format.format(new StringReader(sw.toString()), writer);
                } else {
                    // No format specified, keep original behavior
                    MavenXpp3Writer mavenXpp3Writer = new MavenXpp3Writer();
                    mavenXpp3Writer.write(writer, model);
                }
            } catch (IOException e) {
                throw new UncheckedIOException("Could not write POM file: " + pom, e);
            }
        } else {
            Document document;
            try {
                document = new SAXBuilder().build(pom.toFile());
            } catch (JDOMException e) {
                throw new RuntimeException("Could not parse POM file: " + pom, e);
            } catch (IOException e) {
                throw new UncheckedIOException("Could not read POM file: " + pom, e);
            }
            String indentation = format != null && format.getIndent() != null ? format.getIndent() : XMLFormat.findIndentation(pom);
            try (Writer writer = writerSupplier.get()) {
                MavenJDOMWriter mavenJDOMWriter = new MavenJDOMWriter(indentation);
                XMLOutputter xmlOutputter = format != null ? format.createXmlOutputter()
                        : XMLFormat.DEFAULT.createXmlOutputter();
                mavenJDOMWriter.write(model, document, writer, xmlOutputter);
            } catch (IOException e) {
                throw new UncheckedIOException("Could not write to Writer", e);
            }
        }
    }
}
