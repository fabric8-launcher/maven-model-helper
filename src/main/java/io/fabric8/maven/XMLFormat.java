package io.fabric8.maven;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.Reader;
import java.io.UncheckedIOException;
import java.io.Writer;
import java.nio.file.Files;
import java.nio.file.Path;

import org.jdom2.Document;
import org.jdom2.JDOMException;
import org.jdom2.input.SAXBuilder;
import org.jdom2.output.Format;
import org.jdom2.output.LineSeparator;
import org.jdom2.output.XMLOutputter;

/**
 * Output format for an XML document
 * <p>
 * This experimental class is a simple wrapper around JDOM2's {@link XMLOutputter} to provide a simple way to format XML
 */
public class XMLFormat {

    public static final XMLFormat DEFAULT = XMLFormat.builder().textMode(TextMode.PRESERVE).build();

    private final String indent;

    private final boolean insertLineBreakBetweenMajorSections;

    private final TextMode textMode;

    private final String lineSeparator;

    private XMLFormat(Builder builder) {
        this.indent = builder.indent;
        this.insertLineBreakBetweenMajorSections = builder.insertLineBreakBetweenMajorSections;
        this.textMode = builder.textMode;
        this.lineSeparator = builder.lineSeparator;
    }

    /**
     * @return the number of spaces to use for indentation
     */
    public String getIndent() {
        return indent;
    }

    /**
     * @return true if a line break should be inserted between major sections
     */
    public boolean isInsertLineBreakBetweenMajorSections() {
        return insertLineBreakBetweenMajorSections;
    }

    public TextMode getTextMode() {
        return textMode;
    }

    /**
     * Format the XML from the given reader
     *
     * @param reader the reader
     * @return the formatted XML
     */
    public String format(Reader reader) {
        Document document;
        try {
            document = new SAXBuilder().build(reader);
            XMLOutputter xmlOutputter = createXmlOutputter();
            return xmlOutputter.outputString(document);
        } catch (JDOMException e) {
            throw new RuntimeException("Could not parse XML", e);
        } catch (IOException e) {
            throw new UncheckedIOException("Could not read XML", e);
        }
    }

    public void format(Reader reader, Writer writer) {
        Document document;
        try {
            document = new SAXBuilder().build(reader);
            XMLOutputter xmlOutputter = createXmlOutputter();
            xmlOutputter.output(document, writer);
        } catch (JDOMException e) {
            throw new RuntimeException("Could not parse XML", e);
        } catch (IOException e) {
            throw new UncheckedIOException("Could not read XML", e);
        }
    }

    XMLOutputter createXmlOutputter() {
        XMLOutputter xmlOutputter = new XMLOutputter();
        Format format = Format.getRawFormat();
        format.setIndent(indent);
        format.setLineSeparator(lineSeparator);
        format.setTextMode(Format.TextMode.valueOf(textMode.name()));
        if (insertLineBreakBetweenMajorSections) {
            // Insert line breaks between major sections
            xmlOutputter.setXMLOutputProcessor(new LineBreakProcessor());
        }
        xmlOutputter.setFormat(format);
        return xmlOutputter;
    }

    /**
     * Find the indentation used in the POM file
     *
     * @param pom the path to the POM file
     * @return the indentation used in the POM file
     */
    static String findIndentation(Path pom) {
        try (BufferedReader br = Files.newBufferedReader(pom)) {
            String line;
            while ((line = br.readLine()) != null) {
                int idx = line.indexOf("<");
                // We don't care about the first line or unindented lines
                if (idx > 0) {
                    return line.substring(0, idx);
                }
            }
        } catch (IOException e) {
            throw new UncheckedIOException("Could not read POM file: " + pom, e);
        }
        return XMLFormat.DEFAULT.getIndent();
    }

    public enum TextMode {
        /**
         * Mode for literal text preservation.
         */
        PRESERVE,

        /**
         * Mode for text trimming (left and right trim).
         */
        TRIM,

        /**
         * Mode for text normalization (left and right trim plus internal
         * whitespace is normalized to a single space.
         *
         * @see org.jdom2.Element#getTextNormalize
         */
        NORMALIZE,

        /**
         * Mode for text trimming of content consisting of nothing but
         * whitespace but otherwise not changing output.
         */
        TRIM_FULL_WHITE;
    }

    /**
     * Create a new builder
     *
     * @return a new builder
     */
    public static Builder builder() {
        return new Builder();
    }

    /**
     * Create a new builder with the given {@link XMLFormat}
     *
     * @param format the format
     * @return a new builder
     */
    public static Builder builder(XMLFormat format) {
        return new Builder()
                .indent(format.getIndent())
                .insertLineBreakBetweenMajorSections(format.isInsertLineBreakBetweenMajorSections())
                .textMode(format.getTextMode());
    }

    public static class Builder {
        private String indent = "  ";

        private boolean insertLineBreakBetweenMajorSections = false;

        private TextMode textMode = TextMode.TRIM;

        private String lineSeparator = LineSeparator.UNIX.value();

        Builder() {
        }

        public Builder indent(String indent) {
            this.indent = indent;
            return this;
        }

        public Builder insertLineBreakBetweenMajorSections() {
            return insertLineBreakBetweenMajorSections(true);
        }

        public Builder insertLineBreakBetweenMajorSections(boolean insertLineBreakBetweenMajorSections) {
            this.insertLineBreakBetweenMajorSections = insertLineBreakBetweenMajorSections;
            return this;
        }

        public Builder textMode(TextMode textMode) {
            this.textMode = textMode;
            return this;
        }

        public Builder lineSeparator(String lineSeparator) {
            this.lineSeparator = lineSeparator;
            return this;
        }

        public XMLFormat build() {
            return new XMLFormat(this);
        }
    }
}
