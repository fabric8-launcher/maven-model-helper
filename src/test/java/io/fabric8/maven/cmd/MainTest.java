package io.fabric8.maven.cmd;

import static org.assertj.core.api.Assertions.assertThat;

import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

import org.apache.maven.model.Model;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import io.fabric8.maven.Maven;

/**
 * @author <a href="mailto:ggastald@redhat.com">George Gastaldi</a>
 */
class MainTest {

    @Test
    void should_merge_both_poms() throws Exception {
        Path target = Paths.get(getClass().getResource("target-pom.xml").toURI());
        Path source = Paths.get(getClass().getResource("source-pom.xml").toURI());
        String[] args = {
                "merge",
                target.toString(),
                source.toString()
        };
        Main.main(args);
        Approvals.verify(target.toFile());
    }

    @Test
    void should_keep_ordered_properties_last_on_merge(@TempDir Path tmpDir) throws Exception {
        Path target = Paths.get(getClass().getResource("properties/target-pom.xml").toURI());
        Path source = Paths.get(getClass().getResource("properties/source-pom.xml").toURI());

        Path newTarget = tmpDir.resolve("target-pom.xml");
        Files.copy(target, newTarget);

        String[] args = {
                "merge",
                newTarget.toString(),
                source.toString()
        };
        Main.main(args);
        Approvals.verify(newTarget.toFile());
    }

    @Test
    void should_change_project_metadata(@TempDir Path tempDir) {
        Path target = tempDir.resolve("foo.pom");
        Maven.writeModel(new Model(), target);
        String[] args = {
                "update-metadata",
                target.toString(),
                "my-name",
                "my-description"
        };
        Main.main(args);
        Model model = Maven.readModel(target);
        assertThat(model.getName()).isEqualTo("my-name");
        assertThat(model.getDescription()).isEqualTo("my-description");
    }
}
