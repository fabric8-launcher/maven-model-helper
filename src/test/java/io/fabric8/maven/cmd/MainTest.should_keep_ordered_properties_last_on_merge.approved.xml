<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <groupId>io.fabric8</groupId>
    <artifactId>foo</artifactId>
    <version>1</version>
    <properties>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
        <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
        <maven.compiler.source>{java.version}</maven.compiler.source>
        <maven.compiler.target>{java.version}</maven.compiler.target>
        <maven.compiler.parameters>true</maven.compiler.parameters>
        <!-- Platform: Begin -->
        <quarkus.platform.group-id>{quarkus.platform.group-id}</quarkus.platform.group-id>
        <quarkus.platform.artifact-id>{quarkus.platform.artifact-id}</quarkus.platform.artifact-id>
        <quarkus.platform.version>{quarkus.platform.version}</quarkus.platform.version>
        <!-- Platform: End -->
        <quarkus-plugin.version>{quarkus.maven-plugin.version}</quarkus-plugin.version>
        <compiler-plugin.version>{maven-compiler-plugin.version}</compiler-plugin.version>
        <surefire-plugin.version>{maven-surefire-plugin.version}</surefire-plugin.version>
        <foo>bar</foo>
        <fruit>pineapple</fruit>
    </properties>
</project>
